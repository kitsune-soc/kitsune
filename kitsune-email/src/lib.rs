use askama::Template;
use mrml::{mjml::Mjml, prelude::render::Options as RenderOptions};
use thiserror::Error;
use typed_builder::TypedBuilder;

type Result<T, E = Error> = std::result::Result<T, E>;

#[derive(Debug, Error)]
pub enum Error {
    #[error(transparent)]
    Templating(#[from] askama::Error),

    #[error(transparent)]
    RenderParsing(#[from] mrml::prelude::parse::Error),

    #[error(transparent)]
    Rendering(#[from] mrml::prelude::render::Error),
}

#[derive(Template, TypedBuilder)]
#[template(escape = "html", path = "verify.mjml")]
struct VerifyMail<'a> {
    domain: &'a str,
    username: &'a str,
    verify_link: &'a str,
}

impl VerifyMail<'_> {
    fn render_email(&self) -> Result<String> {
        let rendered_mjml = self.render()?;
        let parsed_mjml = Mjml::parse(rendered_mjml)?;

        parsed_mjml
            .render(&RenderOptions::default())
            .map_err(Error::from)
    }
}

#[cfg(test)]
mod test {
    use crate::VerifyMail;
    use insta::assert_snapshot;

    #[test]
    fn test_render() {
        let mail = VerifyMail {
            domain: "citadel-station.example",
            username: "shodan",
            verify_link: "https://citadel-station.example/verify/perfect-immortal-machine",
        };
        assert_snapshot!(mail.render_email().unwrap());
    }
}
