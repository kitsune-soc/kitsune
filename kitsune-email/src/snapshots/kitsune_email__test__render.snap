---
source: kitsune-email/src/lib.rs
expression: mail.render_email().unwrap()
---
<!doctype html><html xmlns="http://www.w3.org/1999/xhtml" xmlns:v="urn:schemas-microsoft-com:vml" xmlns:o="urn:schemas-microsoft-com:office:office"><head><title></title><!--[if !mso]><!--><meta http-equiv="X-UA-Compatible" content="IE=edge"><!--<![endif]--><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta name="viewport" content="width=device-width, initial-scale=1">
<style type="text/css">
#outlook a { padding: 0; }
body { margin: 0; padding: 0; -webkit-text-size-adjust: 100%; -ms-text-size-adjust: 100%; }
table, td { border-collapse: collapse; mso-table-lspace: 0pt; mso-table-rspace: 0pt; }
img { border: 0; height: auto; line-height: 100%; outline: none; text-decoration: none; -ms-interpolation-mode: bicubic; }
p { display: block; margin: 13px 0; }
</style>
<!--[if mso]>
<noscript>
<xml>
<o:OfficeDocumentSettings>
  <o:AllowPNG/>
  <o:PixelsPerInch>96</o:PixelsPerInch>
</o:OfficeDocumentSettings>
</xml>
</noscript>
<![endif]-->
<!--[if lte mso 11]>
<style type="text/css">
.mj-outlook-group-fix { width:100% !important; }
</style>
<![endif]-->
<style type="text/css">@media only screen and (min-width:480px) { .mj-column-per-100 { width:100% !important; max-width:100%; }  }</style><style media="screen and (min-width:480px)">.moz-text-html .mj-column-per-100 { width:100% !important; max-width:100%; } </style></head><body style="word-spacing:normal;"><div><!--[if mso | IE]><table border="0" cellpadding="0" cellspacing="0" role="presentation" align="center" width="600" style="width:600px;"><tr><td style="line-height:0px;font-size:0px;mso-line-height-rule:exactly;"><![endif]--><div style="margin:0px auto;max-width:600px;"><table border="0" cellpadding="0" cellspacing="0" role="presentation" align="center" style="width:100%;"><tbody><tr><td style="direction:ltr;font-size:0px;padding:20px 0;text-align:center;"><!--[if mso | IE]><table border="0" cellpadding="0" cellspacing="0" role="presentation"><![endif]--><!--[if mso | IE]><tr><![endif]--><!--[if mso | IE]><td style="vertical-align:top;width:600px;"><![endif]--><div class="mj-outlook-group-fix mj-column-per-100" style="font-size:0px;text-align:left;direction:ltr;display:inline-block;vertical-align:top;width:100%;"><table border="0" cellpadding="0" cellspacing="0" role="presentation" width="100%" style="vertical-align:top;"><tbody><tr><td align="center" style="font-size:0px;padding:10px 25px;word-break:break-word;"><div style="font-family:monospace;font-size:30px;line-height:17px;text-align:center;color:#000000;">
          Kitsune
        </div></td></tr><tr><td align="center" style="font-size:0px;padding:10px 25px;word-break:break-word;"><p style="border-top:solid 4px #000000;font-size:1px;margin:0px auto;width:100%;"></p><!--[if mso | IE]><table border="0" cellpadding="0" cellspacing="0" role="presentation" align="center" width="550px" style="border-top:solid 4px #000000;font-size:1px;margin:0px auto;width:550px;"><tr><td style="height:0;line-height:0;">&nbsp;</td></tr></table><![endif]--></td></tr><tr><td align="left" style="font-size:0px;padding:10px 25px;word-break:break-word;"><div style="font-family:monospace;font-size:25px;line-height:17px;text-align:left;color:#000000;">
          Verify your account
        </div></td></tr><tr><td align="left" style="font-size:0px;padding:10px 25px;word-break:break-word;"><div style="font-family:monospace;font-size:13px;line-height:17px;text-align:left;color:#000000;">
          Congratulations to your new account on citadel-station.example!
          <p>Verify your email address to unlock your account (@shodan)!</p></div></td></tr><tr><td align="center" vertical-align="middle" style="font-size:0px;padding:10px 25px;word-break:break-word;"><table border="0" cellpadding="0" cellspacing="0" role="presentation" style="border-collapse:separate;line-height:100%;"><tbody><tr><td align="center" bgcolor="#414141" role="presentation" valign="middle" style="border:none;border-radius:3px;cursor:auto;mso-padding-alt:10px 25px;background:#414141;"><a href="https://citadel-station.example/verify/perfect-immortal-machine" target="_blank" style="display:inline-block;background:#414141;color:#ffffff;font-family:monospace;font-size:13px;font-weight:normal;line-height:17px;margin:0;text-decoration:none;text-transform:none;padding:10px 25px;mso-padding-alt:0px;border-radius:3px;">
          Verify my account!
        </a></td></tr></tbody></table></td></tr><tr><td align="left" style="font-size:0px;padding:10px 25px;word-break:break-word;"><div style="font-family:monospace;font-size:13px;line-height:17px;text-align:left;color:#000000;">
          You didn't register an account? Feel free to just ignore this email!
        </div></td></tr><tr><td align="center" style="font-size:0px;padding:10px 25px;word-break:break-word;"><div style="font-family:monospace;font-size:10px;line-height:17px;text-align:center;color:#000000;">
          citadel-station.example powered by <a style="text-decoration:none;" href="https://joinkitsune.org">Kitsune</a></div></td></tr></tbody></table></div><!--[if mso | IE]></td><![endif]--><!--[if mso | IE]></tr><![endif]--><!--[if mso | IE]></table><![endif]--></td></tr></tbody></table></div><!--[if mso | IE]></td></tr></table><![endif]--></div></body></html>
